doc: Add production app documentation (#51)
TheThingsIndustries/generic-node-docs: doc/content/applications/se-vanilla/uplinks.md

---
title: "uplinks"
description: ""
weight: 1
---

> The vanilla app uplinks are compromised of two simple messages (heartbeat and temperature/humidity) to reduce the complexity and show a simple example of the device functionality.

Reviewer: ```suggestion
The vanilla app uplinks are composed of two simple messages, a heartbeat and a temperature/humidity uplink, to reduce the complexity and show a simple example of the device functionality.
```

## Payload formatters

Please see [The Things Stack Javascript payload formatter documentation](https://www.thethingsindustries.com/docs/integrations/payload-formatters/javascript/) for information on how to use payload formatters.

> The javascript payload formatter below can be used to decode the vanilla app.

Reviewer: ```suggestion
The Javascript payload formatter below can be used to decode the vanilla app.
```

```javascript

> function decodeUplink(input) {

Reviewer 2: Can be formatted to:
```javascript
function decodeUplink(input) {
    var data = {};
    if (input.fPort == 1) {
        data.batteryVoltage = (input.bytes[0] / 10);
        data.fwVersion = input.bytes[1] + (input.bytes[2] / 10);
    }
    if (input.fPort == 2) {
        data.temperature = ((input.bytes[0] << 8) + input.bytes[1]) / 10;
        data.humidity = ((input.bytes[2] << 8) + input.bytes[3]) / 10;
    }
    return {
        data: data,
    };
}
```

Maybe worth adding to the Device Repository instead ? For the old Uno/Nodes, we allow people to select one of the default sketches as a firmware version, so maybe we could have the production app as one of the available firmwares, and it could use this formatter.

Author: With the current design of our GNSE repo, I don't recommend having a production app as standalone app.

The reason for this is that the applications were designed to allow for flexibility and versatility.
Adding a production app there will actively diminish the other applications purpose.

With sensors_lorawan and secure_element_lorawan I think users have all the foundation they need to build something similar to the production app.

I would like to also highlight that this this device addresses a different segment of developers than the ones using UNOs and The Things Node.

Author: > Maybe worth adding to the Device Repository instead ?

I think it should be in both

  var data = {};
  if (input.fPort == 1)
  {
  data.BATTERY_VOLTAGE = (input.bytes[0]/10);
  data.GNSE_FW_VERSION = input.bytes[1] + (input.bytes[2]/10);
  }

  if (input.fPort == 2)
  {
  data.TEMPERATURE = ((input.bytes[0] << 8) + input.bytes[1])/10;
  data.HUMIDITY = ((input.bytes[2] << 8) + input.bytes[3])/10;
  }
  return {
    data: data,
  };
}
```

## Heartbeat

The heartbeat uplink message on FPort 1 contains:
- The device battery voltage (useful for replacing the batteries)
- The firmware version (useful for checking which commands to send)

> 

Reviewer: ```suggestion
```

## Temperature and Humidity
The temperature and humidity uplink message on FPort 2 contains:

> - The temperature in Celsius.

Reviewer 2: Try to be consistent with the ending dots.

Author: Fixed this one, can't find others

> - The humidity in percentage (between 0% to 100%)

Reviewer: ```suggestion
- The relative humidity percentage (between 0% to 100%)
```
